src/midgard/polyline2.cc

```diff
@@ -42,17 +44,209 @@ typename coord_t::value_type Polyline2<coord_t>::Length(const container_t& pts)
+    auto d = line_segment.DistanceSquared(c, tmp);
+    if (d > dmax) {
+      dmax = d;
+      hfidx = idx;
+    }
+  }
+
+  // If (dmax < epsilon_sq) then we have a relatively straight line between (start,end).
+  // A standard Douglas-Peucker algorithm would immediately decimate all the points
+  // between (start,end). In this modified version, we use our tiled-point-space to
+  // determine if decimating the line would result in a self-intersection.
+  //
+  // We use our tiled space to determine the points along the "epsilon buffer zone" of
+  // the line (start,end). Because our tiled-point-space is coarse, our
+  // "get_points_near_segment" query will contain points both of interest and not.
+  // Consider this amazing ascii art example:
+  //
+  //                i             k
+  //                 \           /
+  //                  \         /
+  //                   \       /
+  //                    \     /
+  //   s - - - - - - - - - - - - - - - - - - - - - - - - - - - - e
+  //     `  .             \ /                             `
+  //            `  .       j                 .
+  //                  `  c        `
+  //
+  // s=start, e=end. c is a point along the polyline between s & e. We are considering
+  // getting rid of c because it is within epsilon of (a,b).
+  //
+  // All the points shown in this hypothetical example are returned from the call to
+  // "get_points_near_segment".
+  //
+  // As you can see, a completely separate portion of our polygon (i, j, k) would
+  // self-intersect if we simplified. To detect this, we perform a triangle
+  // containment test of point j using the triangle (s, c, e), see that its contained,
+  // and decide not to simplify. While this example only has one point c between
+  // (a,b), there is typically more than one. The logic below will create a triangle
+  // using every point c between start and end and perform containment tests for all
+  // "nearby" points for every (start,c,end) triangle. We can stop as soon as we find
+  // an unexpected point inside our triangle.
+  if (dmax < epsilon_sq) {
+    // This returns the points in the "epsilon buffer zone" along the line (start, end).
+    std::unordered_set<size_t> line_buffer_points;
+    point_tile_index.get_points_near_segment(LineSegment2<PointLL>(start, end), line_buffer_points);
+
+    // We only care about checking for triangle containment for points that are not
+    // along the polyline [start,end] - so we can remove those straightaway.
+    for (size_t i = sidx; i <= eidx; i++) {
+      line_buffer_points.erase(i);
+    }
+
+    bool can_simplify = true;
+    for (size_t cidx = sidx + 1; (cidx < eidx) && can_simplify; cidx++) {
+      const PointLL& c = point_tile_index.points[cidx];
+
+      for (size_t point_idx : line_buffer_points) {
+        const PointLL& p = point_tile_index.points[point_idx];
+        if (triangle_contains(start, c, end, p)) {
+          can_simplify = false;
+#if 0
```

Comment:
So be `#ifdef LOGGING_LEVEL_TRACE` no?

Reply:
After thinking about it, I've decided that `#if 0` block was just for debugging, so I've removed it.  thx.